File: software/examples/delphi/ExampleRemoteCallback.pas
program ExampleRemoteCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletRemoteSwitchV2;

type
  TExample = class
  private
    ipcon: TIPConnection;
    rs: TBrickletRemoteSwitchV2;
  public
    procedure RemoteStatusACB(sender: TBrickletRemoteSwitchV2; const houseCode: byte;
                              const receiverCode: byte; const switchTo: byte;
                              const repeats: word);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change XYZ to the UID of your Remote Switch Bricklet 2.0 }

var
  e: TExample;

{ Callback procedure for remote status a callback }
procedure TExample.RemoteStatusACB(sender: TBrickletRemoteSwitchV2; const houseCode: byte;
                                   const receiverCode: byte; const switchTo: byte;
                                   const repeats: word);
begin
  WriteLn(Format('House Code: %d', [houseCode]));
  WriteLn(Format('Receiver Code: %d', [receiverCode]));
  WriteLn(Format('Switch To: %d', [switchTo]));
  WriteLn(Format('Repeats: %d', [repeats]));
  WriteLn('');
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  rs := TBrickletRemoteSwitchV2.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register remote status a callback to procedure RemoteStatusACB }
  rs.OnRemoteStatusA := {$ifdef FPC}@{$endif}RemoteStatusACB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleRemoteCallback.vb
Imports System
Imports Tinkerforge

Module ExampleRemoteCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your Remote Switch Bricklet 2.0

    ' Callback subroutine for remote status a callback
    Sub RemoteStatusACB(ByVal sender As BrickletRemoteSwitchV2, ByVal houseCode As Byte, _
                        ByVal receiverCode As Byte, ByVal switchTo As Byte, _
                        ByVal repeats As Integer)
        Console.WriteLine("House Code: " + houseCode.ToString())
        Console.WriteLine("Receiver Code: " + receiverCode.ToString())

        If switchTo = BrickletRemoteSwitchV2.SWITCH_TO_OFF Then
            Console.WriteLine("Switch To: Off")
        Else If switchTo = BrickletRemoteSwitchV2.SWITCH_TO_ON Then
            Console.WriteLine("Switch To: On")
        End If

        Console.WriteLine("Repeats: " + repeats.ToString())
        Console.WriteLine("")
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim rs As New BrickletRemoteSwitchV2(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Configure to receive from remote type A with minimum repeats set to 1 and enable callback
        rs.SetRemoteConfiguration(BrickletRemoteSwitchV2.REMOTE_TYPE_A, 1, True)

        ' Register remote status a callback to subroutine RemoteStatusACB
        AddHandler rs.RemoteStatusACallback, AddressOf RemoteStatusACB

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
